lib/api/completeMultipartUpload.js

```diff
@@ -499,31 +504,44 @@ function completeMultipartUpload(authInfo, request, log, callback) {
                                 }
                                 return next(null, mpuBucket, keysToDelete,
                                     aggregateETag, extraPartLocations,
-                                    destinationBucket, generatedVersionId);
+                                    destinationBucket, generatedVersionId,
+                                    droppedMPUSize);
                             });
                     }
                     return next(null, mpuBucket, keysToDelete, aggregateETag,
                         extraPartLocations, destinationBucket,
-                        generatedVersionId);
+                        generatedVersionId, droppedMPUSize);
                 });
         },
         function deletePartsMetadata(mpuBucket, keysToDelete, aggregateETag,
-            extraPartLocations, destinationBucket, generatedVersionId, next) {
+            extraPartLocations, destinationBucket, generatedVersionId, droppedMPUSize, next) {
             services.batchDeleteObjectMetadata(mpuBucket.getName(),
                 keysToDelete, log, err => next(err, extraPartLocations,
-                    destinationBucket, aggregateETag, generatedVersionId));
+                    destinationBucket, aggregateETag, generatedVersionId, droppedMPUSize));
         },
         function batchDeleteExtraParts(extraPartLocations, destinationBucket,
-            aggregateETag, generatedVersionId, next) {
+            aggregateETag, generatedVersionId, droppedMPUSize, next) {
             if (extraPartLocations && extraPartLocations.length > 0) {
-                return data.batchDelete(extraPartLocations, request.method,
-                    null, log, err => {
-                        if (err) {
-                            return next(err);
-                        }
-                        return next(null, destinationBucket, aggregateETag,
-                            generatedVersionId);
+                return data.batchDelete(extraPartLocations, request.method, null, log, err => {
+                    if (err) {
+                        return next(err);
+                    }
+
```

Comment:
"droppedMPUSize" can be computed around here (sum of every extraPartLocation's size), as it is the place where we drop the extra parts : code will be must simpler & easier to follow, with no risk of mistake between all the steps...

this also makes it easier to handle partial delete object metadata errors : i.e. update quota with only the parts which were succesfully deleted...

Reply:
I see you resolved the comment but just to confirm: the extraPartLocations do not have the information about the size of each part. We need to compute it before, so it is easier to process all parts at once to avoid processing the dropped parts only.